### PR TITLE
Farm: replace in environment values requires a value

### PR DESCRIPTION
**Pull request by kalisp**
## Changelog Description
This PR provides way how to skip or update values for particular environment variables only for farm rendering.

## Additional info
Environment variables on artist machines might be different than on worker machines, maybe pointing to `demo` license on artist machines and `real` license on workers etc.

## Testing notes:
1. configure profile
`ayon+settings://core/filter_farm_environment`
image
(I added temporary variables with `Project environments` field for easier testing.)
3. publish on farm, open log of render job and look for :
    `Skipping` for skipped env vars
    value of modified env var a bit lower when env vars are set from exported `.json` file

### CODE REVIEW

File: server/settings/main.py [hunk 1]
```diff
@@ -169,6 +169,46 @@ class VersionStartCategoryModel(BaseSettingsModel):
...
+        title="Task types",
+        enum_resolver=task_types_enum
+    )
+
+    task_names: list[str] = SettingsField(
+        default_factory=list,
+        title="Task names"
+    )
+
+    folders: list[str] = SettingsField(
```
Comment (iLLiCiTiT) [inline server/settings/main.py:197]: I still don't think this is good idea...

File: server/settings/main.py [hunk 2]
```diff
@@ -169,6 +169,46 @@ class VersionStartCategoryModel(BaseSettingsModel):
...
+class EnvironmentReplacementModel(BaseSettingsModel):
+    environment_key: str = SettingsField("", title="Enviroment variable")
+    pattern: str = SettingsField("", title="Pattern")
+    replacement: str = SettingsField("", title="Replacement")
+
+
+class FilterFarmEnvironmentModel(BaseSettingsModel):
+    _layout = "expanded"
+
+    hosts: list[str] = SettingsField(
```
Comment (iLLiCiTiT) [inline server/settings/main.py:181]:
```suggestion
    host_names: list[str] = SettingsField(
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            project_settings["core"]["filter_farm_environment"])
+
+        if not filter_env_profiles:
+            self.log.debug("No profiles found for env var filtering")
+            return
+
+        task_entity = data["task_entity"]
+
+        filter_data = {
+            "hosts": self.host_name,
```
Comment (iLLiCiTiT) [inline client/ayon_core/hooks/pre_filter_farm_environments.py:37]:
```suggestion
            "host_names": self.host_name,
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            "task_types": task_entity["taskType"],
+            "tasks": task_entity["name"],
```
Comment (iLLiCiTiT) [inline client/ayon_core/hooks/pre_filter_farm_environments.py:39]:
```suggestion
            "task_names": task_entity["name"],
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    def _modify_environment_variables(self, calculated_env, matching_profile):
+        """Modify environment variable values."""
+        for env_item in matching_profile["replace_in_environment"]:
+            value = calculated_env.get(env_item["environment_key"])
+            if not value:
+                continue
+
+            value = re.sub(value, env_item["pattern"], env_item["replacement"])
+            calculated_env[env_item["environment_key"]] = value
```
Comment (iLLiCiTiT) [inline client/ayon_core/hooks/pre_filter_farm_environments.py:68]:
```suggestion
            key = env_item["environment_key"]
            value = calculated_env.get(key)
            if not value:
                continue

            value = re.sub(value, env_item["pattern"], env_item["replacement"])
            if value:
                calculated_env[key] = value
            else:
                calculated_env.pop(key)
            
```

File: server/settings/main.py [hunk 6]
```diff
@@ -169,6 +169,46 @@ class VersionStartCategoryModel(BaseSettingsModel):
...
+        default_factory=list,
+        title="Task names"
+    )
+
+    folders: list[str] = SettingsField(
+        default_factory=list,
+        title="Folders"
+    )
+
+    skip_environment: list[str] = SettingsField(
```
Comment (iLLiCiTiT) [inline server/settings/main.py:202]:
```suggestion
    skip_env_keys: list[str] = SettingsField(
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+        calculated_env = copy.deepcopy(self.launch_context.env)
+
+        calculated_env = self._skip_environment_variables(
+            calculated_env, matching_profile)
+
+        calculated_env = self._modify_environment_variables(
+            calculated_env, matching_profile)
+
+        self.launch_context.env = calculated_env
```
Comment (iLLiCiTiT) [inline client/ayon_core/hooks/pre_filter_farm_environments.py:58]: You could do just:
```suggestion
        self._skip_environment_variables(
            self.launch_context.env, matching_profile)

        self._modify_environment_variables(
            self.launch_context.env, matching_profile)
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py (hunk 5 above)
Comment (kalisp) [reply client/ayon_core/hooks/pre_filter_farm_environments.py:68]: I was thinking about it and I am not sure this is right. But I am not sure, that current implementation (eg. there could be FOO with empty value) is.

Comment (iLLiCiTiT) [reply client/ayon_core/hooks/pre_filter_farm_environments.py:68]: Empty value means that the env should not be set.

File: server/settings/main.py (hunk 1 above)
Comment (kalisp) [reply server/settings/main.py]: Changed to `folder_paths`. 

Maybe you want to modify different env vars for `shots` vs `characters`?

File: client/ayon_core/hooks/pre_filter_farm_environments.py (hunk 3 above)
Comment (kalisp) [reply client/ayon_core/hooks/pre_filter_farm_environments.py]: Updated.

File: client/ayon_core/hooks/pre_filter_farm_environments.py (hunk 5 above)
Comment (kalisp) [reply client/ayon_core/hooks/pre_filter_farm_environments.py:63]: What about use case where you are using env var placeholders in file paths? 
(path_to_licence: `//home/blabla/license{LICENSE_VARIANT}` )

Comment (iLLiCiTiT) [reply client/ayon_core/hooks/pre_filter_farm_environments.py:63]: Not sure how is that connected. If output of
```python
value = re.sub(value, env_item["pattern"], env_item["replacement"])
```
is empty string, it means the environment variable is empty.

If you want to replace value in `"//home/blabla/license{LICENSE_VARIANT}"` you'll probably replace only `"{LICENSE_VARIANT}"`, in worst case scenario the value will be ``"//home/blabla/license"`` which is not `""`.

**Review by MilaKudr — APPROVED** [review]

File: client/ayon_core/hooks/pre_filter_farm_environments.py (hunk 5 above)
Comment (kalisp) [reply client/ayon_core/hooks/pre_filter_farm_environments.py]: Ok, updated.

File: server/settings/main.py [hunk 8]
```diff
@@ -219,6 +259,9 @@ class CoreSettings(BaseSettingsModel):
         title="Project environments",
         section="---"
     )
+    filter_farm_environment: list[FilterFarmEnvironmentModel] = SettingsField(
```
Comment (iLLiCiTiT) [inline server/settings/main.py:262]: Just one last thing to point. I didn't realize these are actually profiles...
```suggestion
    filter_farm_profiles: list[FilterFarmEnvironmentModel] = SettingsField(
```

File: server/settings/main.py [hunk 9]
```diff
@@ -313,5 +356,6 @@ def validate_json(cls, value):
     "project_environments": json.dumps(
         {},
         indent=4
-    )
+    ),
+    "filter_farm_environment": [],
```
Comment (iLLiCiTiT) [inline server/settings/main.py:360]:
```suggestion
    "filter_farm_profiles": [],
```

File: client/ayon_core/hooks/pre_filter_farm_environments.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    order = 1000
+
+    launch_types = {LaunchTypes.farm_publish}
+
+    def execute(self):
+        data = self.launch_context.data
+        project_settings = data["project_settings"]
+        filter_env_profiles = (
+            project_settings["core"]["filter_farm_environment"])
```
Comment (iLLiCiTiT) [inline client/ayon_core/hooks/pre_filter_farm_environments.py:26]:
```suggestion
            project_settings["core"]["filter_farm_profiles"])
```

File: server/settings/main.py [hunk 11]
```diff
@@ -169,6 +169,46 @@ class VersionStartCategoryModel(BaseSettingsModel):
     )
 
 
+class EnvironmentReplacementModel(BaseSettingsModel):
+    environment_key: str = SettingsField("", title="Enviroment variable")
+    pattern: str = SettingsField("", title="Pattern")
+    replacement: str = SettingsField("", title="Replacement")
+
+
+class FilterFarmEnvironmentModel(BaseSettingsModel):
```
Comment (iLLiCiTiT) [inline server/settings/main.py:178]:
```suggestion
class FilterEnvsProfileModel(BaseSettingsModel):
```

**Review by iLLiCiTiT — APPROVED** [review]

**Review by 64qam — APPROVED** [review]

**Review by 64qam — APPROVED** [review]

**Pull request merged**